subprojects/docs/src/samples/userguide/dependencyManagement/customizingResolution/selectionRule/kotlin/build.gradle.kts

```diff
@@ -0,0 +1,171 @@
+repositories {
+    ivy {
+        url = uri("file://${projectDir}/repo")
+    }
+}
+
+// Set up the status scheme so that "experimental" is a valid status for "org.sample" artifacts
+
+class StatusRule : ComponentMetadataRule {
+
+    override fun execute(componentMetadataContext: ComponentMetadataContext) {
+        val details = componentMetadataContext.details
+        if (details.id.group == "org.sample") {
+            details.statusScheme = listOf("experimental", "integration", "milestone", "release")
+        }
+    }
+}
+dependencies {
+    components {
+        all(StatusRule::class.java)
+    }
+}
+
+// tag::reject-version-1-1[]
+configurations {
+    create("rejectConfig") {
+        resolutionStrategy {
+            componentSelection {
+                // Accept the highest version matching the requested version that isn't '1.5'
+                all {
+                    if (candidate.group == "org.sample" && candidate.module == "api" && candidate.version == "1.5") {
+                        reject("version 1.5 is broken for 'org.sample:api'")
+                    }
+                }
+            }
+        }
+    }
+}
+
+dependencies {
+    "rejectConfig"("org.sample:api:1.+")
+}
+// end::reject-version-1-1[]
+
+task("printRejectConfig") {
+    doLast {
+        configurations["rejectConfig"].forEach { println("Resolved: ${it.name}") }
+    }
+}
+
+// tag::component-selection-with-metadata[]
+configurations {
+    create("metadataRulesConfig") {
+        resolutionStrategy {
+            componentSelection {
+                // Reject any versions with a status of 'experimental'
+                // TODO how to pass a closure receiving two arguments here? The API needs to  be more type-safe
+                /*
+                all { ComponentSelection selection, ComponentMetadata metadata ->
+                    if (selection.candidate.group == 'org.sample' && metadata.status == 'experimental') {
+                        selection.reject("don't use experimental candidates from 'org.sample'")
+                    }
+                }
+                // Accept the highest version with either a "release" branch or a status of 'milestone'
+                withModule("org.sample:api") { ComponentSelection selection, IvyModuleDescriptor descriptor, ComponentMetadata metadata ->
+                    if (descriptor.branch != "release" && metadata.status != 'milestone') {
+                        selection.reject("'org.sample:api' must have testing branch or milestone status")
+                    }
+                }
+                */
```

Comment:
@jnizet, good point, `ComponentSelectionRule` is clearly missing type safe members. Could you please open an issue on `gradle/gradle` ?

Reply:
See #6607